src/sbuf.c

```diff
@@ -938,56 +951,100 @@ static void setup_tls(struct tls_config *conf, const char *pfx, int sslmode,
 			tls_config_verify_client_optional(conf);
 		}
 	}
+
+	return true;
 }
 
-void sbuf_tls_setup(void)
+bool sbuf_tls_setup(void)
 {
 	int err;
+	bool ok;
+	struct tls_config *new_client_accept_conf = NULL;
+	struct tls_config *new_server_connect_conf = NULL;
+	struct tls *new_client_accept_base = NULL;
 
 	if (cf_client_tls_sslmode != SSLMODE_DISABLED) {
-		if (!*cf_client_tls_key_file || !*cf_client_tls_cert_file)
-			die("To allow TLS connections from clients, client_tls_key_file and client_tls_cert_file must be set.");
+		if (!*cf_client_tls_key_file || !*cf_client_tls_cert_file) {
+			log_error("To allow TLS connections from clients, client_tls_key_file and client_tls_cert_file must be set.");
+			return false;
+		}
 	}
 	if (cf_auth_type == AUTH_CERT) {
-		if (cf_client_tls_sslmode != SSLMODE_VERIFY_FULL)
-			die("auth_type=cert requires client_tls_sslmode=SSLMODE_VERIFY_FULL");
-		if (*cf_client_tls_ca_file == '\0')
-			die("auth_type=cert requires client_tls_ca_file");
+		if (cf_client_tls_sslmode != SSLMODE_VERIFY_FULL) {
+			log_error("auth_type=cert requires client_tls_sslmode=SSLMODE_VERIFY_FULL");
+			return false;
+		}
+		if (*cf_client_tls_ca_file == '\0') {
+			log_error("auth_type=cert requires client_tls_ca_file");
+			return false;
+		}
 	} else if (cf_client_tls_sslmode > SSLMODE_VERIFY_CA && *cf_client_tls_ca_file == '\0') {
-		die("client_tls_sslmode requires client_tls_ca_file");
+		log_error("client_tls_sslmode requires client_tls_ca_file");
+		return false;
 	}
 
 	err = tls_init();
 	if (err)
 		fatal("tls_init failed");
 
 	if (cf_server_tls_sslmode != SSLMODE_DISABLED) {
-		server_connect_conf = tls_config_new();
-		if (!server_connect_conf)
-			die("tls_config_new failed 1");
-		setup_tls(server_connect_conf, "server_tls", cf_server_tls_sslmode,
+		new_server_connect_conf = tls_config_new();
+		if (!new_server_connect_conf) {
+			log_error("tls_config_new failed 1");
+			return false;
+		}
+		ok = setup_tls(new_server_connect_conf, "server_tls", cf_server_tls_sslmode,
 			  cf_server_tls_protocols, cf_server_tls_ciphers,
 			  cf_server_tls_key_file, cf_server_tls_cert_file,
 			  cf_server_tls_ca_file, "", "", true);
+		if (!ok) {
+			tls_config_free(new_server_connect_conf);
+			return false;
+		}
 	}
 
 	if (cf_client_tls_sslmode != SSLMODE_DISABLED) {
-		client_accept_conf = tls_config_new();
-		if (!client_accept_conf)
-			die("tls_config_new failed 2");
-		setup_tls(client_accept_conf, "client_tls", cf_client_tls_sslmode,
+		new_client_accept_conf = tls_config_new();
+		if (!new_client_accept_conf) {
+			log_error("tls_config_new failed 2");
+			tls_config_free(new_client_accept_conf);
```

Comment:
Let's collect all these free calls at the end of the function behind a "failed" label and just `goto failed` here. There are many examples of that approach in the pgbouncer code.